src/plugins/discover/public/application/main/services/discover_app_state_container.ts

```diff
@@ -7,13 +7,64 @@
  */
 
 import {
+  createStateContainer,
   createStateContainerReactHelpers,
   ReduxLikeStateContainer,
 } from '@kbn/kibana-utils-plugin/common';
-import { AggregateQuery, Filter, Query } from '@kbn/es-query';
-import { VIEW_MODE } from '@kbn/saved-search-plugin/public';
+import { AggregateQuery, Filter, FilterStateStore, Query } from '@kbn/es-query';
+import { SavedSearch, VIEW_MODE } from '@kbn/saved-search-plugin/public';
+import { IKbnUrlStateStorage, ISyncStateRef, syncState } from '@kbn/kibana-utils-plugin/public';
+import { cloneDeep, isEqual } from 'lodash';
+import { connectToQueryState, syncGlobalQueryStateWithUrl } from '@kbn/data-plugin/public';
+import { DiscoverServices } from '../../../build_services';
+import { addLog } from '../../../utils/add_log';
+import { getValidFilters } from '../../../utils/get_valid_filters';
+import { cleanupUrlState } from '../utils/cleanup_url_state';
+import { getStateDefaults } from '../utils/get_state_defaults';
+import { handleSourceColumnState } from '../../../utils/state_helpers';
+import { APP_STATE_URL_KEY, AppStateUrl, isEqualState, setState } from './discover_state';
 import { DiscoverGridSettings } from '../../../components/discover_grid/types';
 
+export interface DiscoverAppStateContainer extends ReduxLikeStateContainer<AppState> {
+  /**
+   * Returns the previous state, used for diffing e.g. if fetching new data is necessary
+   */
+  getPrevious: () => AppState;
+  /**
+   * Determines if the current state is different from the initial state
+   */
+  hasChanged: () => boolean;
+  /**
+   * Initializes the state by the given saved search and starts syncing the state with the URL
+   * @param currentSavedSearch
+   */
+  initAndSync: (currentSavedSearch: SavedSearch) => () => void;
+  /**
+   * Resets the state by the given saved search
+   * @param savedSearch
+   */
+  resetBySavedSearch: (savedSearch: SavedSearch) => void;
+  /**
+   * Resets the current state to the initial state
+   */
+  resetInitialState: () => void;
+  /**
+   * Replaces the given state in the URL, instead of pushing a new history entry
+   * @param newState
+   */
+  replaceUrlState: (newState: AppState) => void;
```

Comment:
Is it necessary to expose `replaceUrlState` when `update({ ... }, true)` seems to do the same thing?

Reply:
also true, thx! replaced